### DISCUSSION TITLE
Neo CoolCam PIR Sensor Config Issue

### DISCUSSION THREAD

**Discussion by bwosborne2** [zwave-js/zwave-js#2734] [General] [answered]
Neo Coolcam PIR Sensor https://devices.zwave-js.io/?jumpTo=0x0258:0x0003:0x0083:0.0
Configuration parameter 6 - LED Blink Enable  starts with an illegal default value of 5 (should be 1) and cannot be set to 0 to disable blink.

There appears to be no useful information from a users perspective about how to troubleshoot or report / discuss issues. I chanced upon this and hope it is the correct place.

I am testing on moving from openHAB and have been involved somewhat supporting the binding there. Their documentation clearly states how to collect necessary logs and where to report the issue.

**Comment by AlCalzone** [comment, marked as answer]
> There appears to be no useful information from a users perspective about how to troubleshoot or report / discuss issues. I chanced upon this and hope it is the correct place.

As a developer who's on Github daily I might be routine blinded here, but that's what the "Issues" tab is for. There, we have bug report templates for actual bugs or missing configuration files. And those templates guide you through the process and guide you a bit how to get the information we need.

If you have any suggestions on how to make this more obvious, I'll be glad to take your input. Keep in mind that `node-zwave-js` is used by multiple projects which have their own documentation and "workflows".

**Reply by bwosborne2** [reply]
> Keep in mind that `node-zwave-js` is used by multiple projects which have their own documentation and "workflows".

Keep in mind many users of your library are not developers familiar with GitHub. Personally, this is the first time I even noticed Discussions on GitHub.

I was a little hesitant to even post here because I am using zwavejs->zwavejs2mqtt->HomeAssistant zwavejs2mqtt Hass.io integration.

I can open an issue on Github. Perhaps something should be mentioned in  the Troubleshooting section of the documentation?

This is just a suggestion for your consideration. I am not submitting a PR for that.

**Reply by AlCalzone** [reply]
> Keep in mind many users of your library are not developers familiar with GitHub.

And that's fine. Discussions is a beta feature we have early access for (I'm not sure if its out of beta yet). I wish we could pin a note there guiding users to the issues tab.

> I can open an issue on Github.

Yes please.

> I was a little hesitant to even post here because I am using zwavejs->zwavejs2mqtt->HomeAssistant zwavejs2mqtt Hass.io integration.

To be fair, there are a lot of moving parts and it is not always obvious where a problem comes from. We usually send users to the repo for their application first before opening one here - but in this case it seems pretty clear.

> Perhaps something should be mentioned in the Troubleshooting section of the documentation?
This is just a suggestion for your consideration. I am not submitting a PR for that.

No problem - it does make sense to mention that somewhere. However the process of collecting logs already differs between applications and has changed a couple of times with updates to HA and zwavejs2mqtt.

**Reply by bwosborne2** [reply]
I have opened an Issue on this. Since I am currently still testing, getting back to HA after 2 years of openHAB, I have more flexibility to test.

-- New topic, but this is my thread ;) --

By the way, the OH database on openSmartHouse has a REST API access. Would it be useful at all to try and export devices not currently in your database?

**Reply by AlCalzone** [reply]
Sorry, I missed the comment. We do have an import script that can use the OpenSmartHouse DB as a source. This can be used to create starter files, but they usually need to be cleaned up manually because our checks and style guide are stricter than for the other Z-Wave projects.
https://zwave-js.github.io/node-zwave-js/#/config-files/importing-from-others?id=opensmarthouse-z-wave-device-database-openhab

**Reply by bwosborne2** [reply]
I also talked with the developer.  He is rather open to data sharing. I have worked with him some on database management and a little bit of PHP programming.

I know about 6 years ago he offered to share with OZW and was refused. He still has a menu item that exports to the old OZW config.

Do you have an API or just the GitHub source? OSH is in a MySQL database, periodically exported as XML to Github.

**Reply by AlCalzone** [reply]
We only have the import script on Github. Our config updates are more of a pull workflow than a push workflow mainly because of the checks I mentioned above and the danger of breaking manually cleaned up files.

@blhoward2 and I have discussed regularly pulling **new configs** from the Z-Wave alliance website. We could probably consider the OpenSmartHouse DB as another source since not every device ends up on the alliance site. This should also be limited to devices we don't have yet. This would be as simple as knowing which IDs to import and run the script with them.

Do you want to ask the OSH dev to join in on this discussion?

**Reply by blhoward2** [reply]
I'm actually not even sure if the OSH code works anymore.

**Reply by bwosborne2** [reply]
Currently when somebody makes a change on OSH they ask for it to be reviewed and an email gets sent. After somebody approves the entry, there is another email. I currently get those emails. I see I cannot tag Chris here so I will ask him to join the discussion. I do not know where you live. I am in the US and Chris is in New Zealand so time zones could be a factor.

I just sent Chris a link to this discussion.

**Reply by bwosborne2** [reply]
> I'm actually not even sure if the OSH code works anymore.

You mean the script? I might be able to give a look. I actually helped by writing the export scripts used by OSH to openHAB.

**Reply by blhoward2** [reply]
Yes, we've made substantial improvements over time to preserve comments and such but only to the zwa section. That required changing how we stringify the json, how we add and remove elements to the object to reorder them, etc. We've also removed some previously mandatory fields. It would require testing and we'd need to be absolutely sure any change anywhere outside that function didn't break something else as OSH is at best a secondary source of files.

**Reply by bwosborne2** [reply]
I'll take a look.

Shortly after I started with OpenHAB 2 years ago I spent 4 months going through their device database cleaning up a bunch of content issues.

FYI, this api call gets the whole database.

https://www.opensmarthouse.org/dmxConnect/api/zwavedatabase/device/list.php?filter=&manufacturer=-1&limit=-1

 It can also be filtered, but it is easier than browsing all the web pages.

**Reply by blhoward2** [reply]
> Currently when somebody makes a change on OSH they ask for it to be reviewed and an email gets sent. After somebody approves the entry, there is another email. I currently get those emails. I see I cannot tag Chris here so I will ask him to join the discussion. I do not know where you live. I am in the US and Chris is in New Zealand so time zones could be a factor.
> 
> 
> 
> I just sent Chris a link to this discussion.

Ideally our goal is to move away from this model, though, as we've found lots of errors in all of the major repositories, not to mention inconsistencies in labels, etc. We're setting up to import from the zwave alliance somewhat automatically as new devices get added. And we've already begun working with several of the most popular manufacturers on our platform (and hopefully more in the future) to add files either pre-release or right after release for new devices. 

So user submitted changes will hopefully (fingers crossed) be limited to where devices aren't certified or listed by the zwa. Hopefully users having to submit corrections will eventually be a thing of the past. We're also working to template and re-verify what we have for devices over time so the same parameter is described the same for every device of that manufacturer.

We've done some mass imports from ZWA but it's labor intensive because we aren't merging it as it comes in. We are trying to impose standards on labels and such so every file needs touched. We realized we were getting bogged down in devices unlikely to ever see the platform so we shifted gears to working on the popular ones and templating and adding them as we go, so we don't get have all of the ZWA ones. For most users a missing device is trivial in that it says unknown but for newer devices we setup the parameters and such from the device itself so they still work fine. 

If you want to see what is in OSH that's not also in ZWA and merge the delta by all means go for it. I can send you a pile of jsons for every device listed in the zwa database to build an index from and check against.

**Reply by bwosborne2** [reply]
> Ideally our goal is to move away from this model, though, as we've found lots of errors in all of the major repositories, not to mention inconsistencies in labels, etc. We're setting up to import from the zwave alliance somewhat automatically as new devices get added. And we've already begun working with several of the most popular manufacturers on our platform (and hopefully more in the future) to add files either pre-release or right after release for new devices.
> 
> So user submitted changes will hopefully (fingers crossed) be limited to where devices aren't certified or listed by the zwa. Hopefully users having to submit corrections will eventually be a thing of the past. We're also working to template and re-verify what we have for devices over time so the same parameter is described the same for every device of that manufacturer.

It appears your style guide requires some manual review to even attempt to meet your style guide requirements. 

> We've done some mass imports from ZWA but it's labor intensive because we aren't merging it as it comes in. We are trying to impose standards on labels and such so every file needs touched. We realized we were getting bogged down in devices unlikely to ever see the platform so we shifted gears to working on the popular ones and templating and adding them as we go, so we don't get have all of the ZWA ones. For most users a missing device is trivial in that it says unknown but for newer devices we setup the parameters and such from the device itself so they still work fine.
> 

Automatically importing from ZWA or even vendors will still involve manual work to meet the style guidelines for labels, as an example. I have also found many manuals from the vendor use horribly broken English to document configuration parameters.

An automated job COULD pull in the data, build a proposed JSON file and open a PR but you would also possibly need the referenced source to sanely edit the file to style guidelines.

Something similar could be done with a user form. OSH entries all include the device manual that hopefully has the Z-Wave informaiton.

> If you want to see what is in OSH that's not also in ZWA and merge the delta by all means go for it. I can send you a pile of jsons for every device listed in the zwa database to build an index from and check against.

 I could build that from GitHub, correct? I assume there is no automated way to search against your database?I have not looked closely at how your web based browser functions.

**Reply by AlCalzone** [reply]
> I assume there is no automated way to search against your database?

There kinda is, but the documentation currently lies in the private repo for the device browser. You can verify if a single configuration exists with a GET request to
```
https://devices.zwave-js.io/public_api/getdeviceinfo/<ManufacturerID>/<ProductType>/<ProductID>/<FirmwareVersion>
```
but that endpoint is currently rate-limited. The easiest would probably to use the `@zwave-js/config` package and work with the device index directly: https://zwave-js.github.io/node-zwave-js/#/api/config-manager?id=getindex

> An automated job COULD pull in the data, build a proposed JSON file and open a PR

For single files, I can trigger our bot to do that with the comment `@zwave-js-bot import config <device-id> from <source>`. That will create a PR which contributors with write access can edit further. For mass imports that isn't practical though.

> I have also found many manuals from the vendor use horribly broken English to document configuration parameters.

Yeah, I don't see how we're ever going to get around manual review. That's the cost of wanting to provide top-notch device configurations.

> Something similar could be done with a user form. OSH entries all include the device manual that hopefully has the Z-Wave informaiton.

That would be the ideal UX, so nobody has to bother with editing the JSON files directly. We could even include some stylistic checks in there, reducing the need for manual review.
But to be honest, I don't have the time for that, so this would depend on someone from the community stepping up.

**Reply by bwosborne2** [reply]
> That would be the ideal UX, so nobody has to bother with editing the JSON files directly. We could even include some stylistic checks in there, reducing the need for manual review.
> But to be honest, I don't have the time for that, so this would depend on someone from the community stepping up.
Well, I am busy too but seem to be directed more & more into web work, Here is an asciidoc based documentation mockup, using Antora, that I put together for OSH, which is/was going to be a fork of openHAB.

http://bosborne.freeshell.org/antora/developers/index.php

It appears all your web is currently hosted on Github and I have not really worked much with GitHub Pages.

**Reply by AlCalzone** [reply]
Only the static stuff - i.e. documentation. Everything that needs a DB and/or secret access is hosted on Heroku.
